src/server/frontend_xwayland/xcb_connection.cpp

```diff
@@ -69,6 +69,21 @@ auto number_to_readable_name(uint32_t number) -> std::string
     ss << ends[number % length_of(ends)];
     return ss.str();
 }
+
+auto connect_to_fd(int fd) -> xcb_connection_t*
+{
+    xcb_connection_t* connection = xcb_connect_to_fd(fd, nullptr);
+    if (xcb_connection_has_error(connection))
+    {
+        xcb_disconnect(connection);
+        close(fd);
+        BOOST_THROW_EXCEPTION(std::runtime_error("xcb_connect_to_fd() failed"));
```

Comment:
It would probably be helpful for future debugging to distinguish between the various forms of error [this can return](https://xcb.freedesktop.org/manual/group__XCB__Core__API.html#ga70a6bade94bd2824db552abcf5fbdbe3).

Non-blocking, though.